### PR TITLE
Add methods to the trait error handling that will return the type/schema of the generated error column and their aggregation: #91

### LINKED ISSUE #91: Add methods to the `trait ErrorHandling` that will return the type/schema of the generated `ErrorColumn` and their aggregation.
## Background
There's a lack of information for the caller about the actual implementation of the `trait ErrorHandling`. The information might be useful for whatever reason - call optimization etc.

## Feature
Add methods to the `trait ErrorHandling` that will return the `DataType` of the returned `ErrorColumn` and of the aggregation (that should be probably `Option[DataType]` in case the aggregation actually does not add a column.

**Advanced:**
Try to find a solution that `class ErrorColumn` would be able to return the attached Column type too (while keeping it still a value-class).

### PR DESCRIPTION
**Pull request by TebaleloS**
* Added `errorColumnType` method in ErrorHandling trait
* Added  'errorColumnAggregationType` in the Error Handling trait
* Implement `errorColumntype` in the `ErrorMessageArray` class and in the `ErrorHandlingFilterRowsWithErrors`
* Implement `errorColumnAggregationType` in the `ErrorMessageArray` class and in the `ErrorHandlingFilterRowsWithErrors`

Closes #91

### CODE REVIEW

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala [hunk 1]
```diff
@@ -66,6 +68,21 @@ case class ErrorMessageArray(errorColumnName: String = ErrorMessageArray.default
     dataFrame.withColumnIfDoesNotExist(joinToExisting)(errorColumnName, reMap(aggregatedWithoutNulls))
   }
 
+  override def errorColumnType: DataType = {
+    decomposeMap(col(errorColumnName)).expr.dataType
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala:72]: I actually thought that this is a good idea, at first, but I played with it on Spark Shell and I don't think that what you are trying to achieve here is possible. @benedeki please feel free to disagree.

The reason is that you are creating a new Column object here, a generic one, not associated with any dataframe so no internal structure / metadata will be associated with it. So you are only creating a placeholder for a column. 

The `decomposeMap(col(errorColumnName))` make sense, but it will produce a Column, not a dataType. Then, `expr` will produce a general expression full of NamePlaceholder elements, and then `.dataType` call will fail on: `org.apache.spark.SparkUnsupportedOperationException: Cannot evaluate expression: NamePlaceholder`

These are deeper Spark internals and I don't know if you can hack it so that you get a schema in `dataType` out of it.

---

Why not to take a simpler approach? The structure is already defined above. Check how the err column is being added into the DataFrame, line:

`dataFrame.withColumn(errorColName, reMap(array_union(deMap(col(errorColName)), colToUnion)))`

or `dataFrame.withColumnIfDoesNotExist(joinToExisting)(errorColumnName, reMap(aggregatedWithoutNulls))` and try to understand the internals of functions involved in these computations.

Basically it's in the `recomposeMap` - maybe you can somehow extract & reuse that, but maybe it would be good enough to just code the structure manually into your `errorColumnType` according to this

Comment (lsulak) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala:72]: Btw, in terms of the schema that you are looking for, every concrete implementation should know what schema the err column will be used. See https://github.com/AbsaOSS/spark-commons/pull/103, the documentation is I think outstanding. The function `createErrorAsColumn` is responsible for creating a error Column expression. If you check what's inside, it is using `transformErrorSubmitToColumn`. It might be a good idea to read the documentation carefully and confirm your understanding with concrete implementations and unit tests. 

For example, `ErrorMessageArray` is extending `ErrorHandling` and `TransformIntoErrorMessage`. In the `TransformIntoErrorMessage`, you can see the definition of `transformErrorSubmitToColumn` that I mentioned above - this function is important for you - and the schema is defined there :) 

David's changes simplify things. You might be a bit confused about the function naming, the master branch that you are using still has `putErrorToColumn` (not transformErrorSubmitToColumn) and `evaluate` function (instead of `transformErrorSubmitToColumn`) but the concept is the same. So in your case `evaluate` function contains the err Column structure.

Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala:72]: Okay, thank you  Ladislav. This is crucial and insightful, and I will be sure to ask if there is something that confuses me.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 2]
```diff
@@ -87,4 +88,18 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
 
     assert(result == expected)
   }
+
+  test("errorColumnType should return a BooleanType") {
```
Comment (benedeki) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:92]: These tests only kind of duplicate the implementation. I would try to add one errorcolumn's (`ErrorColumn.column`) column to an (empty) dataframe and check it's structure.
And in the other case apply the aggregation function to an the empty DataFrame and do the same.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala [hunk 3]
```diff
@@ -36,4 +37,8 @@ trait ErrorHandling {
       .getOrElse(ErrorMessageSubmitWithoutColumn(errType, errCode, errMessage, additionalInfo))
     putErrorToColumn(toSubmit)
   }
+
+  def errorColumnType: DataType
+
+  def errorColumnAggregationType: Option[DataType]
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala:43]: David and I talked (in one of his previous PRs I think) that it can be nice to put abstract methods/functions into upper parts of a given class/trait. Would you mind just moving it to the top / more up?

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrors.scala [hunk 4]
```diff
@@ -18,11 +18,12 @@ package za.co.absa.spark.commons.errorhandling.implementations
 
 import org.apache.spark.sql.{Column, DataFrame}
 import org.apache.spark.sql.functions.{coalesce, lit}
+import org.apache.spark.sql.types.{BooleanType, DataType}
 import za.co.absa.spark.commons.errorhandling.ErrorMessageSubmit
 import za.co.absa.spark.commons.errorhandling.partials.ErrorHandlingCommon
 
 /**
- * Class implement the functionality of filtering rows with columns.
+ * Class implements the functionality of filtering rows witch has some error (any of the error columns is not NULL).
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrors.scala:26]: typo

```suggestion
 * Class implements the functionality of filtering rows that have some error (any of the error columns is not NULL).
```

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrors.scala [hunk 5]
```diff
@@ -46,4 +47,13 @@ object ErrorHandlingFilterRowsWithErrors extends ErrorHandlingCommon {
     dataFrame.filter(!coalesce(columns: _*))
   }
 
+  /**
+   * @return BooleanType of Datatype object
+   */
+  val errorColumnType: DataType = BooleanType
+
+  /**
+   * @return None since no column added during the aggregation
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrors.scala:56]: a potential improvement 

```suggestion
   * @return None since no error-related column added during the aggregation
```

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala [hunk 6]
```diff
@@ -20,7 +20,8 @@ import org.apache.spark.sql.catalyst.expressions.{CaseWhen, Expression}
 import org.apache.spark.sql.{Column, DataFrame}
 import za.co.absa.spark.commons.errorhandling.{ErrorHandling, ErrorMessageSubmit}
 import za.co.absa.spark.commons.errorhandling.types._
-import org.apache.spark.sql.functions.when
+import org.apache.spark.sql.functions.{when}
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala:23]:
```suggestion
import org.apache.spark.sql.functions.when
```

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala (hunk 3 above)
Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala:43]: I will try to.

Comment (lsulak) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala:43]: yes, the most 'abstract' one. But I don't know if you agree or disagree with my comment...?

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 7]
```diff
@@ -17,19 +17,14 @@
 package za.co.absa.spark.commons.errorhandling.implementations
 
 import org.apache.spark.sql.DataFrame
-import org.scalatest.funsuite.AnyFunSuite
-import za.co.absa.spark.commons.errorhandling.types.ErrorWhen
-import za.co.absa.spark.commons.test.SparkTestBase
 import org.apache.spark.sql.functions.{col, length}
+import org.apache.spark.sql.types.{ArrayType, LongType, MapType, NullType, StringType, StructField, StructType}
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:21]: `NullType` is not used

Comment (TebaleloS) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala]: Removed

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala (hunk 3 above)
Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala:43]: I agree with your comment.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala (hunk 2 above)
Comment (benedeki) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:92]:
```scala
    val emptyDf = spark.emptyDataFrame
    val errorColumn: ErrorColumn = ErrorHandlingFilterRowsWithErrors.putErrorToColumn(???) // just add some error submission)
    val testDf = emptyDf.withColumn('error_column', errorColumn.column)
    // assert testDf schema to the ErrorHandlingFilterRowsWithErrors.errorColumnType
   // similarl with errorColumnAggregationType - well compare that the schema of the empty Df and testDf did not change after putErrorsWithGrouping
```

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala [hunk 8]
```diff
@@ -21,6 +21,7 @@ import org.apache.spark.sql.{Column, DataFrame}
 import za.co.absa.spark.commons.errorhandling.{ErrorHandling, ErrorMessageSubmit}
 import za.co.absa.spark.commons.errorhandling.types._
 import org.apache.spark.sql.functions.when
+import org.apache.spark.sql.types.DataType
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala:24]: this is never used

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 9]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+      "1st error", 0, "This is an error", Some(errColName)
+    )
+
+    val testDf = emptyDf
+    val expVal = testDf.schema.fields.toList.headOption
+
+    val expectedAfterAgg = ErrorHandlingFilterRowsWithErrors.aggregateErrorColumns(testDf)(errorColumn)
+    val expAggRes = expectedAfterAgg.schema.fields.headOption.headOption
+
+    val results = ErrorHandlingFilterRowsWithErrors.errorColumnAggregationType
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:118]: Just a proposal: `results` -> `actualType`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 10]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
     assert(result == expected)
   }
+
+  test("errorColumnType should return a BooleanType") {
+    val errorColumn: ErrorColumn = ErrorHandlingFilterRowsWithErrors.putErrorToColumn(
+      "Test error 1", 1, "This is a test error", Some(errColName))
+
+    val testDf = emptyDf.withColumn(errColName, errorColumn.column)
+    val expectedResults = testDf.col(errColName).expr.dataType
+    val expval = testDf.schema.fields
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:99]: Just a proposal: `expVal` -> `expectedColumns`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 11]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
 
     assert(result == expected)
   }
+
+  test("errorColumnType should return a BooleanType") {
+    val errorColumn: ErrorColumn = ErrorHandlingFilterRowsWithErrors.putErrorToColumn(
+      "Test error 1", 1, "This is a test error", Some(errColName))
+
+    val testDf = emptyDf.withColumn(errColName, errorColumn.column)
+    val expectedResults = testDf.col(errColName).expr.dataType
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:98]: Just a proposal: `expectedResults` -> `expectedType`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 12]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+    val expval = testDf.schema.fields
+
+    val results = ErrorHandlingFilterRowsWithErrors.errorColumnType
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:101]: Just a proposal: `results` -> `actualType`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 13]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+  test("errorColumnAggregationType should return None since no column is added during the aggregation") {
+    val errorColumn: ErrorColumn = ErrorHandlingFilterRowsWithErrors.putErrorToColumn(
+      "1st error", 0, "This is an error", Some(errColName)
+    )
+
+    val testDf = emptyDf
+    val expVal = testDf.schema.fields.toList.headOption
+
+    val expectedAfterAgg = ErrorHandlingFilterRowsWithErrors.aggregateErrorColumns(testDf)(errorColumn)
+    val expAggRes = expectedAfterAgg.schema.fields.headOption.headOption
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:116]: Remove redundant `.headOption ` (it's there twice)

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 14]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+    assert(results == expectedResults)
+  }
+
+  test("errorColumnAggregationType should return None since no column is added during the aggregation") {
+    val errorColumn: ErrorColumn = ErrorHandlingFilterRowsWithErrors.putErrorToColumn(
+      "1st error", 0, "This is an error", Some(errColName)
+    )
+
+    val testDf = emptyDf
+    val expVal = testDf.schema.fields.toList.headOption
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:113]: Hmm, I'm not sure if this is relevant here. 

Basically you just have an empty list of columns, basically `None`, but that is still from your input DF, but you are trying to test the datatype after the aggregate operation here. I think I would just remove this. 

But please let me know what you think / your reasoning, maybe I'm just not getting something.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 15]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+
+    val expectedAfterAgg = ErrorHandlingFilterRowsWithErrors.aggregateErrorColumns(testDf)(errorColumn)
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:115]: This is not 'expected' but 'actual', right? You are actually performing some computation. So something like `expectedAfterAgg` -> `actualDfAfterAgg`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 16]
```diff
@@ -199,4 +194,34 @@ class ErrorMessageArrayTest extends AnyFunSuite with SparkTestBase {
     assert(result == expected)
 
   }
+
+  test("test return errCol type schema") {
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:198]: These 2 tests are valid and work, but they don't really test any computation. It might be better to do something like what you did with tests in `ErrorHandlingFilterRowsTest`, so to have actual DataFrame and perform actual computation :)

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala (hunk 3 above)
Comment (lsulak) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorHandling.scala:43]: Still there ^, just remaining

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala (hunk 12 above)
Comment (TebaleloS) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:101]: `results` is the actual results that the tested method return. What if we call it `actualResults` or actualType`

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala (hunk 14 above)
Comment (TebaleloS) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:113]: You are right, and I spoke with David and when trying to extract the type of the column added to the DF is not doing what is supposed to. so, I robbed it like that meanwhile I'm doing research about it.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala (hunk 16 above)
Comment (TebaleloS) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:198]: Well, the thing is that the this method returns the hard coded values which are not formed after any computation in `ErrorHandlingFilterRowsWithErrors` class.

Comment (benedeki) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:198]: Use the same logic here as in `ErrorHandlingFilterRowsWithErrorsTest`(_"errorColumnType should return a BooleanType"_).
Add the column to an empty DataFrame, then ask for it and compare to the `errorColumnType`.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala [hunk 17]
```diff
@@ -87,4 +89,35 @@ class ErrorHandlingFilterRowsWithErrorsTest extends AnyFunSuite with SparkTestBa
...
+
+    val testDf = emptyDf
+    val expectedType = testDf.schema.fields.toList.headOption
+
+    val expectedAfterAgg = ErrorHandlingFilterRowsWithErrors.aggregateErrorColumns(testDf)(errorColumn)
+    val expectedTypeAfterAgg = expectedAfterAgg.schema.fields.headOption
+
+    val actualType = ErrorHandlingFilterRowsWithErrors.errorColumnAggregationType
+
+    assert(actualType == expectedType)
```
Comment (benedeki) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorHandlingFilterRowsWithErrorsTest.scala:120]: This assertion is not needed.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 18]
```diff
@@ -199,4 +194,34 @@ class ErrorMessageArrayTest extends AnyFunSuite with SparkTestBase {
...
+    ))
+
+    val results = errorMessageArray.errorColumnType
+    assert(results == expectedResults)
+  }
+
+  test("errorColumnAggregationType should return an ArrayType structure for column added during the aggregation") {
+    val errorMessageArray = ErrorMessageArray("errCol")
+
+    val expectedResults = Some(ArrayType(
```
Comment (benedeki) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:216]: Use the same logic here as in `ErrorHandlingFilterRowsWithErrorsTest` with the only difference that instead of `.headOption` use the field identified by its name ("_errCol_") here.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 19]
```diff
@@ -197,6 +193,41 @@ class ErrorMessageArrayTest extends AnyFunSuite with SparkTestBase {
     val result = resultDfToResult(resultDf)
 
     assert(result == expected)
+  }
+
+  test("errorColumnType should return errCol type schema") {
```
Comment (lsulak) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:198]: I think that this test can be dropped, it doesn't actually do anything apart from copying the same schema as the underlying function `errorColumnType`.

But I like the next test, so I think that you don't need to implement any other tests here in this module :)

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala [hunk 20]
```diff
@@ -417,4 +417,45 @@ object StructTypeImplicits {
...
 
+  implicit class StructTypeComparisonWhileIgnoringNullability(val schema: StructType) {
+    /**
+     * Wraps an array of StructField with StructType
+     *
+     * @param fieldSchema is the actual schema to be wrapped
+     * @return StructType containing the StructFields
+     */
+    def wrapStructFiledWithStructType(fieldSchema: Option[DataType]): Boolean = {
+      val st = StructType(Seq(StructField("errCol", fieldSchema.headOption.head)))
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:428]: Bear in mind that `"errCol"` i.e. name of the err column, can be parametrized as well, see `case class ErrorMessageArray(errorColumnName: String = ErrorMessageArray.defaultErrorColumnName)
`

So perhaps this class or function should also follow similar logic - making it parametrizable, maybe with some default if you want.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala [hunk 21]
```diff
@@ -417,4 +417,45 @@ object StructTypeImplicits {
...
+            false
+          } else {
+            st1.fields.zip(st2.fields).forall {
+              case (f1, f2) =>
+                f1.name == f2.name && dataTypeEqualsIgnoreNullability(f1.dataType, f2.dataType)
+            }
+          }
+
+        case (ArrayType(et1, n1), ArrayType(et2, n2)) =>
+          dataTypeEqualsIgnoreNullability(et1, et2) && n1 == n2
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:452]: I really like this function. :) I think it's useful and I like the way you used recursion and pattern matching. 

However, the code docs says: `while ignoring nullability` and those `n1 == n2` comparisons (here and 1 more below) are comparing `containsNull` condition. 

What did you mean by 'ignoring nullability' then?

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala (hunk 20 above)
[the remark this answers is not in the document]
Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:428]: Yah, that actually work. So I have refactored it.

**Review by lsulak — COMMENTED** [review]
I did another iteration of review and I think that you are nearly there, just a few minor findings

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala (hunk 21 above)
Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:452]: What I mean is the function only compares the field names and their dataTypes. Then it doesn't care if the field cantains null or not.

Comment (lsulak) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:449]: Hmm, but that's my point - it doesn't really only do that, right? So back to my original question, what about those `n1 == n2` ?

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala [hunk 22]
```diff
@@ -417,15 +418,15 @@ object StructTypeImplicits {
     }
   }
 
-  implicit class StructTypeComparisonWhileIgnoringNullability(val schema: StructType) {
+  implicit class StructTypeComparisonWhileIgnoringNullability(val schema: StructType)(implicit errorColumnName: String = ErrorMessageArray.defaultErrorColumnName) {
```
Comment (lsulak) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:417]: I'm not really sure about this implicit here - what if the val for this is named `errColumnName` or `errColName` or something like that? Why not treat it as an ordinary parameter (i.e. non implicit)?

Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:417]: But, with that I was trying to at least get the value of the `errColName` from the ErrorMessageArray.

Comment (TebaleloS) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:417]: I tried to do what you asking here, but unfortunately implicit classes takes only one parameter on their default constructor.
But in stead of parameterizing i just used `val errorColumnName = ErrorMessageArray.defaultErrorColumnName` to get the column name.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala [hunk 23]
```diff
@@ -413,4 +414,45 @@ object StructTypeImplicits {
     }
   }
 
+  implicit class StructTypeComparisonWhileIgnoringNullability(val schema: StructType)(implicit errorColumnName: String = ErrorMessageArray.defaultErrorColumnName) {
```
Comment (benedeki) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala:417]: I like the idea of such a function for the purposes. But would rather use `DataTypeEnhancements.isEquivalentDataType` enhanced to:
```scala
isEquivalentDataType(other: DataType, ignoreNullability: false, ignoreArrayContainsNull: false, ignoreMapValueContainsNull: false): Boolean
```

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 24]
```diff
@@ -197,6 +192,25 @@ class ErrorMessageArrayTest extends AnyFunSuite with SparkTestBase {
     val result = resultDfToResult(resultDf)
 
     assert(result == expected)
+  }
+
+  test("errorColumnAggregationType should return an ArrayType structure for column added during the aggregation") {
+    val errColName = "errCol"
```
Comment (benedeki) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:198]: Just to test, the override works 😉 
```suggestion
    val errColName = "specialErrCol"
```

File: spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala [hunk 25]
```diff
@@ -197,6 +192,25 @@ class ErrorMessageArrayTest extends AnyFunSuite with SparkTestBase {
...
+    val errorSubmitB = ErrorMessageSubmitWithoutColumn("Test error 3", 3, "This is a test error")
+    val e3 = errorMessageArray.putErrorToColumn(errorSubmitB)
+
+    val dfAfterAgg = errorMessageArray.aggregateErrorColumns(srcDf)(e1, e2, e3)
+    val addedColType = dfAfterAgg.select(errColName).schema
+
+    val results = errorMessageArray.errorColumnAggregationType
+    val actualResultsType = DataTypeImplicits.DataTypeEnhancements(addedColType).wrapStructFiledWithStructType(results, errColName)
+
+    assert(actualResultsType)
```
Comment (benedeki) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArrayTest.scala:213]: This would work much better. It is more explanative and test more things.
```suggestion
    val origSchemaSize = srcDf.schema.fieldNames.length
    val dfAfterAgg = errorMessageArray.aggregateErrorColumns(srcDf)(e1, e2, e3)
    assert(dfAfterAgg.schema.fieldNames.length == origSchemaSize +1) // checkc only one field was added...
    assert(dfAfterAgg.schema.fieldNames(origSchemaSize) == errColName) // and is of correct name...
    //... and type
    val addedColType = dfAfterAgg.schema.fields(origSchemaSize).dataType

    // while using `get` in `Option` is discouraged, it's ok here, as it's expected the option to be non-empty; and if empty
    // the test will fail, which is correct
    val result = errorMessageArray.errorColumnAggregationType.get

    import DataTypeImplicits.DataTypeEnhancements
    assert(result.isEquivalentDataType(addedColType))
```

File: spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/DataTypeImplicits.scala [hunk 26]
```diff
@@ -23,6 +23,16 @@ import za.co.absa.spark.commons.implicits.StructTypeImplicits.StructTypeEnhancem
...
+    /**
+     * Wraps an array of StructField with StructType
+     *
+     * @param fieldSchema is the actual schema to be wrapped
+     * @return StructType containing the StructFields
+     */
+    def wrapStructFiledWithStructType(fieldSchema: Option[DataType], errorColumnName: String): Boolean = {
+      val st = StructType(Seq(StructField(errorColumnName, fieldSchema.head)))
+      isEquivalentDataType(st)
+    }
```
Comment (benedeki) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/implicits/DataTypeImplicits.scala:35]: Couple of things wrong with this function (not mentioning it's not needed anymore, so please remove 😄 )
* Function name does not align with the result type (from name one would expect to get back a `StructField`
* `errorColumnName` parameter name is oddly specific to the place where the function was used. These are general functions, they should have a general use, and therefore the parameter names hint its reason in the function.
* `fieldSchema.head` relies on that the `Option` is non-empty. You should never assume that.
*

**Review by benedeki — CHANGES_REQUESTED** [review]
